simplified-main/src/main/java/org/nypl/simplified/main/MainActivity.kt

```diff
@@ -79,49 +81,77 @@ class MainActivity : AppCompatActivity(R.layout.main_host) {
+        .getDynamicLink(intent)
 
-          lateinit var screenID: org.nypl.simplified.deeplinks.controller.api.ScreenID
-          if (deepLink?.getQueryParameter("screen") == "login") {
-            screenID = org.nypl.simplified.deeplinks.controller.api.ScreenID.LOGIN
-          } else {
-            this.logger.error("MainActivity", "Error: screen ID not specified")
-            screenID = org.nypl.simplified.deeplinks.controller.api.ScreenID.UNSPECIFIED
-          }
+    pendingLink.addOnFailureListener(this) { e ->
+      this.logger.error("Failed to retrieve dynamic link: ", e)
+    }
 
-          val barcode = deepLink?.getQueryParameter("barcode")
+    pendingLink.addOnSuccessListener { linkData ->
+      val deepLink = linkData.link
+      if (deepLink == null) {
+        this.logger.error("Pending deep link had no link field")
+        return@addOnSuccessListener
+      }
 
-          val deepLinksController =
-            Services.serviceDirectory()
-              .requireService(org.nypl.simplified.deeplinks.controller.api.DeepLinksControllerType::class.java)
-          deepLinksController.publishDeepLinkEvent(accountID, screenID, barcode)
-        }
+      val libraryID = deepLink.getQueryParameter("libraryid")
+      if (libraryID == null) {
+        this.logger.error("Pending deep link had no libraryid parameter.")
+        return@addOnSuccessListener
+      }
+
+      val barcode = deepLink.getQueryParameter("barcode")
+      if (barcode == null) {
+        this.logger.error("Pending deep link had no barcode parameter.")
+        return@addOnSuccessListener
       }
-      .addOnFailureListener(this) { e ->
-        this.logger.warn("MainActivity", "getDynamicLink:onFailure", e)
+
+      val services =
+        Services.serviceDirectory()
+      val profiles =
+        services.requireService(ProfilesControllerType::class.java)
+      val deepLinksController =
+        services.requireService(DeepLinksControllerType::class.java)
+
+      val accountURI =
+        URI("urn:uuid" + libraryID)
+
+      val accountResult =
+        profiles.profileAccountCreate(accountURI)
+          .get(3L, TimeUnit.MINUTES)
+
+      // XXX: Creating an error report would be nice here.
+      if (accountResult is TaskResult.Failure) {
+        this.logger.error("Unable to create an account with ID {}: ", accountURI)
+        return@addOnSuccessListener
       }
+
+      val accountID =
+        (accountResult as TaskResult.Success).result.id
+      val screenRaw =
+        deepLink.getQueryParameter("screen")
+
+      val screenId =
+        when (screenRaw) {
+          null -> kotlin.run {
+            this.logger.warn("Deep link did not have a screen parameter.")
+            ScreenID.UNSPECIFIED
+          }
+          "login" -> ScreenID.LOGIN
+          else -> kotlin.run {
+            this.logger.warn("Deep link had an unrecognized screen parameter {}.", screenRaw)
+            ScreenID.UNSPECIFIED
```

Comment:
Are these `kotlin.run` needed here? Also, I think it would be better if this "login" value was declared as a `const val` inside the `companion object` block

Reply:
Amusingly, the IDE actually inserted those. I thought it looked a little odd... They're just supposed to be block statements. I think there's an unfortunate syntactic aspect of Kotlin where you intend to declare a block but you actually declare a lambda and then up not executing anything (because the syntax is nearly identical).

I'll remove the `run` calls.

I agree about the constants too. I just copied the original code pretty directly.